applications/FluidDynamicsApplication/custom_processes/weighted_divergence_calculation_process.h

```diff
@@ -0,0 +1,338 @@
+                Vector values_y(number_nodes_element);
+                Vector values_z(number_nodes_element);
+                for(int i_node=0; i_node < number_nodes_element; ++i_node){
+                    values_x[i_node] = r_geometry[i_node].FastGetSolutionStepValue(VELOCITY_X);
+                    values_y[i_node] = r_geometry[i_node].FastGetSolutionStepValue(VELOCITY_Y);
+                    values_z[i_node] = r_geometry[i_node].FastGetSolutionStepValue(VELOCITY_Z);
+                }
+
+                // Set integration points
+                const auto& r_integration_method = r_geometry.GetDefaultIntegrationMethod(); // Default is 0
+                const auto& r_integration_points = r_geometry.IntegrationPoints(r_integration_method); // Default is [3 dimensional integration point(0.333333 , 0.333333 , 0), weight = 0.5]
+                const std::size_t number_of_integration_points = r_integration_points.size(); // Default is 1
+
+                // Set containers of the shape functions and the local gradients
+                const Matrix& rNcontainer = r_geometry.ShapeFunctionsValues(r_integration_method); // [1,3]((0.333333,0.333333,0.333333))
+                const auto& rDN_DeContainer = r_geometry.ShapeFunctionsLocalGradients(r_integration_method); // [1]([3,2]((-1,-1),(1,0),(0,1)))
+
+                // Initialize auxiliary local variables
+                double divergence_current = 0;
+                double velocity_seminorm_current = 0;
+
+                // Loop over integration points
+                for ( IndexType point_number = 0; point_number < number_of_integration_points; ++point_number ){
+                    // Getting the shape functions
+                    noalias(N) = row(rNcontainer, point_number); // [3](0.333333,0.333333,0.333333)
+
+                    // Get the jacobians
+                    GeometryUtils::JacobianOnInitialConfiguration(r_geometry, r_integration_points[point_number], J0);
+                    double detJ0;
+                    Matrix InvJ0;
+                    MathUtils<double>::GeneralizedInvertMatrix(J0, InvJ0, detJ0);
+                    const Matrix& rDN_De = rDN_DeContainer[point_number];
+                    GeometryUtils::ShapeFunctionsGradients(rDN_De, InvJ0, DN_DX);
+
+                    // Compute local gradient
+                    const Vector grad_x = prod(trans(DN_DX), values_x);
+                    const Vector grad_y = prod(trans(DN_DX), values_y);
+                    const Vector grad_z = prod(trans(DN_DX), values_z);
+
+                    // Compute divergence and velocity seminorm
+                    const double aux_current_divergence = grad_x[0] + grad_y[1] + grad_z[2];
+                    const double aux_current_velocity_seminorm = grad_x[0]*grad_x[0] + grad_x[1]*grad_x[1] + grad_x[2]*grad_x[2] + grad_y[0]*grad_y[0] + grad_y[1]*grad_y[1] + grad_y[2]*grad_y[2] + grad_z[0]*grad_z[0] + grad_z[1]*grad_z[1] + grad_z[2]*grad_z[2];
```

Comment:
This is a perfect candidate to be migrated to an `inline` function.

Reply:
and to use the inner_prod function

Reply:
Done.